### PR TITLE
Refactor opers

### PR DESCRIPTION
**Pull request by stevelinton**
This PR refactors a lot of code in opers.c and makes some small changes that improve performance (especially in IS_SUBSET_FLAGS). The main aim is to reduce code duplication and allow the compiler to do the work instead. It depends for performance on a C compiler that will inline and do a LOT of contstant folding.

A major goal was to get all the logic for method selection including the cache in just one place, without repetition, which is now achieved. This should make it easier to experiment with alternative cache layouts and algorithms.

### CODE REVIEW

File: src/opers.c [hunk 1]
```diff
@@ -1821,2769 +1786,612 @@ static inline Obj CacheOper (
...
 #else
 
-#define GET_METHOD_CACHE( oper, i ) \
-    CACHE_OPER( oper, i )
+#define GET_METHOD_CACHE(oper, i) CACHE_OPER(oper, i)
 
 #endif
 
+/* This function actually searches the cache. Normally it should be called
+   with
```
Comment (ChrisJefferson) [inline src/opers.c]: weird formatting

**Comment by fingolfin** [comment]
I'd really like to review this before we merge it, could you please wait a bit with that, so I can return to europe first? :-)

**Comment by stevelinton** [comment]
@fingolfin Happy to wait. In the meantime I've run clang-format on all the new code.

File: src/opers.c [hunk 2]
```diff
@@ -449,48 +449,46 @@ static Int IsSubsetFlagsCalls2;
...
 #endif
                 return False;
             }
         }
-        if ( len2 == 0 ) {
+        if (len2 == 0) {
             return True;
         }
-        if ( len2 < 3 ) {
+        if (len2 < 16) {
```
Comment (fingolfin) [inline src/opers.c]: Why the change from 3 to 16?

File: src/opers.h [hunk 3]
```diff
@@ -235,9 +235,14 @@ extern Obj TRY_NEXT_METHOD;
 **
 **  Note that 'ELM_FLAGS' is a macro, so do not call it  with arguments  that
 **  have side effects.
+**
+**  C_ELM_FLAGS returns a result more convenient for use inside the kernel
+**  since the C compiler can't know that True != False.
```
Comment (fingolfin) [inline src/opers.h]: I don't follow: How is that "more convenient"?
Perhaps you meant to say it enables additional optimization opportunities?

File: src/opers.c [hunk 4]
```diff
@@ -1742,6 +1713,14 @@ static inline Obj TYPE_OBJ_FEO (
     }
 }
 
+/* Method Cache -- we remember recently selected methods in a cache.
+   The effectiveness of this cache is vital for GAP's performance */
+
+
+/* THe next few functions deal with finding and allocating if necessary the cache
```
Comment (fingolfin) [inline src/opers.c]: "THe" -> "The"

File: src/opers.c [hunk 5]
```diff
@@ -1831,788 +1810,340 @@ static inline Obj CacheOper (
...
-    Obj                prec;
 
-    /* try to find an applicable method in the cache                       */
-    prec = INTOBJ_INT(-1);
 
-    do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
+static inline Obj __attribute__((always_inline))
```
Comment (fingolfin) [inline src/opers.c]: `__attribute__((always_inline))` is not portable, so a `configure` check for it, defining a suitable macro, is needed.

File: src/opers.h (hunk 3 above)
Comment (stevelinton) [reply src/opers.h]: Yes. That is what I meant. I'll rephrase.

File: src/opers.c (hunk 2 above)
Comment (stevelinton) [reply src/opers.c]: Experimentally 16 seems to be roughly optimal for modern CPUs and compilers. The gain from pushing it up from 3 to say 10 is quite noticeable. After that the differences are smaller.

Comment (stevelinton) [reply src/opers.c]: Apart from CPU and compiler changes, the total number of flags has increased a lot over the years.

**Comment by stevelinton** [comment]
@fingolfin, yes. A lot of this was driven by examining the assembler output from clang, plus I wanted to experiment with cache size and layout without having to copy the code in  56 times. The cache changes made here seem to help a bit, but there may be cleverer things to do (eg separate the caches for first, second, etc. preference methods entirely).

File: src/opers.c (hunk 2 above)
Comment (fingolfin) [reply src/opers.c]: OK, thanks for the clarification. It's just that this change is kind of hidden amidst the huge refactoring. Might be good idea to mention it in the commit message?

File: src/opers.c [hunk 6]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
-              Bag cacheBag = GET_METHOD_CACHE( oper, 1 );
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[3*STATE(CacheIndex)] = method;
-              cache[3*STATE(CacheIndex)+1] = prec;
-              cache[3*STATE(CacheIndex)+2] = id1;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
+                    if (i > target) {
+                        Obj buf[n + 2];
+                        memmove((void *)buf, (void *)(cache + i),
```
Comment (fingolfin) [inline src/opers.c]: Why use `memmove` here? Since `buf` is allocated above, we know it cannot overlap with `cache`, hence `memcpy` is safe to use, and potentially very slightly faster.

File: src/opers.c [hunk 7]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
-            }
-#ifdef COUNT_OPERS
-          OperationMiss++;
-#endif
+    return method;
+}
+
+static inline void
```
Comment (fingolfin) [inline src/opers.c]: It would be great if there was a comment before this function explaining what it is supposed to do, at least very briefly.

File: src/opers.c [hunk 8]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
+CacheMethod(Obj oper, UInt n, Obj prec, Obj * ids, Obj method)
+{
+
```
Comment (fingolfin) [inline src/opers.c]: Remove this empty line?

File: src/opers.c [hunk 9]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
     prec = INTOBJ_INT(-1);
-
     do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 6 ) );
+        prec = INTOBJ_INT(INT_INTOBJ(prec) + 1);
```
Comment (fingolfin) [inline src/opers.c]: Why not change `prec` to an `Int`? Most places using it seem to just run `INT_INTOBJ` on it anyway; only in `CacheMethod` is the `INTOBJ` value being used (and stored), but (a) only after checking its `Int` value (so it performs a conversion anyway), and (b) that is of course only called once, to record a hit -- so it would seem better to only convert once, instead of constantly converting back and forth, no? (Ah yes, `CallHandleMethodNotFound` also needs the OBJ value, but that, too, is only rarely called).

File: src/opers.c [hunk 10]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
-    return res;
-}
-
-
-/****************************************************************************
-**
-**  DoConstructor4Args( <oper>, <a1>, <a2>, <a3>, <a4> )
-*/
-Obj DoConstructor4Args (
-    Obj
```
Comment (fingolfin) [inline src/opers.c]: why is `0` on its own line in that comment?

File: src/opers.c [hunk 11]
```diff
@@ -5871,8 +3680,45 @@ Obj FuncOPERS_CACHE_INFO (
     SET_ELM_PLIST( list, 7, INTOBJ_INT(IsSubsetFlagsCalls1) );
     SET_ELM_PLIST( list, 8, INTOBJ_INT(IsSubsetFlagsCalls2) );
     SET_ELM_PLIST( list, 9, INTOBJ_INT(OperationNext)       );
+    Obj tensor = NEW_PLIST(IMMUTABLE_TNUM(T_PLIST), CACHE_SIZE);
```
Comment (fingolfin) [inline src/opers.c]: Add a comment explaining what "tensor" is?

File: src/opers.c [hunk 12]
```diff
@@ -5871,8 +3680,45 @@ Obj FuncOPERS_CACHE_INFO (
...
+            for (Int k = 0; k <= 6; k++)
+                SET_ELM_PLIST(
+                    vec, k + 1,
+                    INTOBJ_INT(CacheHitStatistics[i - 1][j - 1][k]));
+        }
+    }    
+    SET_ELM_PLIST(list, 10, tensor);
+    CHANGED_BAG(list);
+
+    Obj mat = NEW_PLIST(IMMUTABLE_TNUM(T_PLIST), CACHE_SIZE+1);
```
Comment (fingolfin) [inline src/opers.c]: ... and also what this is about?

File: src/opers.c [hunk 13]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
+    Obj * cache;
+    Obj   method = 0;
+    UInt  i;
+
+    cache = 1 + ADDR_OBJ(CacheOper(oper, n));
+
+    /* Up to CACHE_SIZE methods might be in the cache */
+    if (prec < INTOBJ_INT(CACHE_SIZE)) {
+        /* This loop runs through those */
+        UInt target = (n + 2) * INT_INTOBJ(prec);
```
Comment (fingolfin) [inline src/opers.c]: Perhaps make this `const` to help the reader realize this (the compiler does that anyway, I guess).

Also, notice how changing `prec` to `Int` simplifies this code. (But of course we can do that later).

File: src/opers.c [hunk 14]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
+                    if (cache[i + j + 2] != ids[j]) {
+                        typematch = 0;
+                        break;
+                    }
+                }
+                if (typematch) {
+                    method = cache[i];
 #ifdef COUNT_OPERS
-              OperationHit++;
+                    CacheHitStatistics[INT_INTOBJ(prec)][i / (n + 2)][n]++;
```
Comment (fingolfin) [inline src/opers.c]: This function almost always uses `n+2`, and only uses `n` twice. So perhaps add an `n+=2;` at the start, and adjust accordingly?

Then `target` would become `const target = n * prec;`

File: src/opers.c [hunk 15]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
-            {
-              Bag cacheBag = GET_METHOD_CACHE( oper, 1 );
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[3*STATE(CacheIndex)] = method;
-              cache[3*STATE(CacheIndex)+1] = prec;
-              cache[3*STATE(CacheIndex)+2] = id1;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
+                    if (i > target) {
+                        Obj buf[n + 2];
```
Comment (fingolfin) [inline src/opers.c]: A comment here which explains what the following three memmove are supposed to would be nice. 

They seem to reorder or swap cache entries?!   cache[i] till cache[i + n + 1]  (a total of n+2 entries) are copied to buf;  then  cache[target] till cache[i-1] (a total of i-target items) are copied over to cache[target+n+2] .  And then finally the content of `buf` are copied to cache[target] till cache[target+n+1].
Aha, and `target` is a multiple of `n+2`, which is important to know to understand this... I think I'd have a much easier time understanding this block of code if `n` was transformed as suggested. (That said, this change, and also that for `prec`, could of course come after your PR is merged).

File: src/opers.c [hunk 16]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
+                        memmove((void *)buf, (void *)(cache + i),
+                                sizeof(Obj) * (n + 2));
+                        memmove((void *)(cache + target + n + 2),
+                                (void *)(cache + target),
+                                sizeof(Obj) * (i - target));
+                        memmove((void *)(cache + target), (void *)buf,
```
Comment (fingolfin) [inline src/opers.c]: This here also could use `memcpy`, couldn't it?

File: src/opers.c (hunk 10 above)
Comment (fingolfin) [inline src/opers.c]: mark these as `static`? That may help the optimizer, too.

File: src/opers.c (hunk 9 above)
Comment (stevelinton) [reply src/opers.c]: It is stored in the cache, which is just a Plist, so what is stored there needs to be legal GAP object. It might be sensible to  convert it immediately on reading from there though.

File: src/opers.c [hunk 17]
```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
...
+    for (UInt i = 0; i < n; i++)
+        cache[2 + i] = ids[i];
+    CHANGED_BAG(cacheBag);
+}
+
+/* These will contain the GAP method selection functions */
+static Obj MethodSelectors[2][7];
+static Obj VerboseMethodSelectors[2][7];
+
+static inline Obj __attribute__((always_inline))
```
Comment (fingolfin) [inline src/opers.c]: I thought I had added that in a comment already, but apparently did not :-/ but I don't really like using a raw `__attribute__` like this -- those should always be wrapped in a macro, and covered by an autoconf check. I just did that work in PR #1779, so once that is merged, you may want to rebase this PR, and start using `ALWAYS_INLINE` instead of `__attribute__((always_inline))`

File: src/opers.c (hunk 9 above)
Comment (fingolfin) [reply src/opers.c]: Of course inside the cache, it is stored as an intobj; but that doesn't mean we have to keep it in that representation while using it. :-)

File: src/opers.c (hunk 17 above)
Comment (stevelinton) [reply src/opers.c]: Thanks, wil do.

Comment (fingolfin) [reply src/opers.c]: PR #1779 has been merged.

File: src/opers.c (hunk 10 above)
Comment (stevelinton) [reply src/opers.c]: They are used elsewhere in the kernel quite a lot. It saves one memory access compared to CALL_4ARGS which is probably not actually a good choice, but making these static now seems like a decision that doesn't belong in this PR.

Comment (fingolfin) [reply src/opers.c]: They are used elsewhere? I am confused, the only file in which the string `DoConstructor4Args` occurs is `opers.c` (of course other files use it indirectly via function pointers).
Anyway, I don't mind either way :)

Comment (stevelinton) [reply src/opers.c]: You're right of course. DoOperation<n>Args for <n> <=3 are used elsewhere though. I'd be inclined to think they should all be static or none of them.

**Comment by stevelinton** [comment]
Does anyone understand why these tests are failing? It looks like there are two versions of methsel1.g etc. around and in some tests one is being compiled and then diffed against the compiled version of the other one.

**Comment by fingolfin** [comment]
When I check out your PR, and do a `make -j8 && make docomp`, it updates `src/c_methsel1.c` and `src/c_type1.c`; and if I then do the same from an HPC-GAP build, it updates `hpcgap/src/c_type1.c`. You'll have to commit these three file to your PR.

File: src/opers.c [hunk 18]
```diff
@@ -1823,61 +1824,25 @@ Obj FuncCOMPACT_TYPE_IDS( Obj self )
...
 **
-*F  DoOperation( <name> ) . . . . . . . . . . . . . . .  make a new operation
+*F  DoOperation<N>Args( <oper>, ... ) . . . . . . . . . . .Operation Handlers
+**
+**  This section of the file provides handlers for operations. The main ones are
+**  DoOperation0Args..DoOperation6Args and the DoVerboseOperation tracing
+**  variants. Then there are variants for constructors. In the following section
+**  are handlers for attributes, properties and the operations related to them.
+**
+**  This code is being refactored (Oct 2017) to reduce repetition. It's efficiency
```
Comment (fingolfin) [inline src/opers.c]: "It's" -> "Its"

File: src/opers.c [hunk 19]
```diff
@@ -1823,61 +1824,25 @@ Obj FuncCOMPACT_TYPE_IDS( Obj self )
...
   NextTypeID = -(1L << NR_SMALL_INT_BITS);
   CallbackForAllBags( FixTypeIDs );
+  CALL_0ARGS(FLUSH_ALL_METHOD_CACHES);
   return INTOBJ_INT(NextTypeID);
 }
 
 /****************************************************************************
 **
-*F  DoOperation( <name> ) . . . . . . . . . . . . . . .  make a new operation
+*F  DoOperation<N>Args( <oper>, ... ) . . . . . . . . . . .Operation Handlers
```
Comment (fingolfin) [inline src/opers.c]: Change ". .Operation" to ".  Operation" (i.e. replace the last period by a space)

File: src/opers.c [hunk 20]
```diff
@@ -1823,61 +1824,25 @@ Obj FuncCOMPACT_TYPE_IDS( Obj self )
...
+**
+**  This section of the file provides handlers for operations. The main ones are
+**  DoOperation0Args..DoOperation6Args and the DoVerboseOperation tracing
+**  variants. Then there are variants for constructors. In the following section
+**  are handlers for attributes, properties and the operations related to them.
+**
+**  This code is being refactored (Oct 2017) to reduce repetition. It's efficiency
+**  now depends on the C compiler inlining some quite large functions and then doing
+**  constant folding to effectively produce a specialised version of the main function
```
Comment (fingolfin) [inline src/opers.c]: (Almost) all of these comments are limited to 78 columns; this one is not; perhaps rewrap?

Or, leave it as is, whatever, I am just glad there is a comment, thanks :-). I merely wanted to point this out, like any other nitpick :-).

File: src/opers.c [hunk 21]
```diff
@@ -1902,10 +1867,18 @@ static inline Obj TYPE_OBJ_FEO (
     }
 }
 
+/* Method Cache -- we remember recently selected methods in a cache.
+   The effectiveness of this cache is vital for GAP's performance */
+
+
+/* The next few functions deal with finding and allocating if necessary the cache 
+   for a given operation and number of arguments, and some locking in HPCGAP */
```
Comment (fingolfin) [inline src/opers.c]: HPCGAP -> HPC-GAP

File: src/opers.c [hunk 22]
```diff
@@ -1981,2769 +1952,627 @@ static inline Obj CacheOper (
...
+static ALWAYS_INLINE Obj GetMethodCached(Obj  oper,
+                                         UInt n,
+                                         Int  prec,
+                                         Obj  ids[])
+{
+    UInt  typematch;
+    Obj * cache;
+    Obj   method = 0;
+    UInt  i;
+    UInt  CacheEntrySize = n + 2;
```
Comment (fingolfin) [inline src/opers.c]: `CacheEntrySize` is a constant, isn't? So perhaps add `const` (the compiler probably won't care, but it helps a human trying to understand the code. I.e. ;-).

Also, since it's a local variable, I'd change it to `cacheEntrySize`.

**Comment by stevelinton** [comment]
@fingolfin Thanks. For some reason make docomp wasn't doing that for me, but it probably just means I need to clean something better. Will dig.

File: src/opers.c [hunk 23]
```diff
@@ -1981,2769 +1952,627 @@ static inline Obj CacheOper (
...
-              arglist[1] = arg2;
-              arglist[2] = arg3;
-              method = CallHandleMethodNotFound(oper, 3, arglist, 0, 1, prec);
-            }
-          
-          /* update the cache */
-          if (method && prec < INTOBJ_INT(CACHE_SIZE))
-            {
-              Bag cacheBag = GET_METHOD_CACHE( oper, 3 );
-
```
Comment (fingolfin) [inline src/opers.c]: This comment is still weirdly formatted - why is 0 on its own line?

Comment (fingolfin) [inline src/opers.c]: Since we are reformatting: perhaps remove those trailing spaces inside the comments (keeping the "correct" is a PITA, and serves no purpose), or even change it to a `//` comment. The same remarks applies elsewhere (I think `clang-format` would also do it, doesn't it?)

File: src/opers.c [hunk 24]
```diff
@@ -6015,35 +3844,72 @@ Obj FuncOPERS_CACHE_INFO (
...
+    SET_ELM_PLIST(list, 8, INTOBJ_INT(IsSubsetFlagsCalls2));
+    SET_ELM_PLIST(list, 9, INTOBJ_INT(OperationNext));
+    SET_ELM_PLIST(list, 10, INTOBJ_INT(WITH_HIDDEN_IMPS_HIT));
+    SET_ELM_PLIST(list, 11, INTOBJ_INT(WITH_HIDDEN_IMPS_MISS));
+    SET_ELM_PLIST(list, 12, INTOBJ_INT(WITH_IMPS_FLAGS_HIT));
+    SET_ELM_PLIST(list, 13, INTOBJ_INT(WITH_IMPS_FLAGS_MISS));
+    
+    /* Now we need to convert the 3d matrix of cache hit counts (by
+       precedence, location found and number of arguments) into a three
+       dimensional GAP matrix (tensor) */
```
Comment (fingolfin) [inline src/opers.c]: Thanks, that comment is really helpful

File: src/opers.c [hunk 25]
```diff
@@ -6015,35 +3844,72 @@ Obj FuncOPERS_CACHE_INFO (
...
+                SET_ELM_PLIST(
+                    vec, k + 1,
+                    INTOBJ_INT(CacheHitStatistics[i - 1][j - 1][k]));
+        }
+    }
+    SET_ELM_PLIST(list, 14, tensor);
+    CHANGED_BAG(list);
+
+    /* and similarly the 2D matrix of cache miss information (by
+       precedence and number of arguments */
```
Comment (fingolfin) [inline src/opers.c]: This is missing a `)`

File: src/system.h [hunk 26]
```diff
@@ -211,11 +211,10 @@ typedef Bag Obj;
 **
 **  'BIPEB' is the  number of bits  per  block, where a  block  fills a UInt,
 **  which must be the same size as a bag identifier.
+**  'LBIPEB' is the log to the base 2 of BIPEB
 **
 */
-enum {
-    BIPEB =  sizeof(UInt) * 8L
-};
+enum { BIPEB = sizeof(UInt) * 8L, LBIPEB = (BIPEB == 64) ? 6L : 5L };
```
Comment (fingolfin) [inline src/system.h]: I'd write this like this:
```C
enum {
    BIPEB = sizeof(UInt) * 8,
    LBIPEB = (BIPEB == 64) ? 6 : 5,
};
```

**Comment by fingolfin** [comment]
@stevelinton huh? No "cleaning" should be necessary? Perhaps I misunderstand?

Anyway, note that it's vital that you do `make && make docomp`, to ensure `docomp` uses a binary built from the current sources -- because the diff we are seeing depend on that.

**Comment by stevelinton** [comment]
@fingolfin Thanks. I think that was the problem. Anyway it seems to be OK now. I'll fix the remaining nitpicks.
